src/Lecture2.hs

```diff
@@ -268,7 +366,21 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval         _ (Lit lit) = Right lit
+eval variables (Var var) = case (lookup var variables) of
+                             Just int -> Right int
+                             Nothing  -> Left (VariableNotFound var)
+eval variables (Add expr1 expr2)
+  | isRight result1 && isRight result2 = Right $ (fromRight' result1 + fromRight' result2)
+  | isLeft result1                     = result1
+  | isLeft result2                     = result2
+  | otherwise                          = error "Computational error"
+    where result1 = eval variables expr1
+          result2 = eval variables expr2
+
+fromRight' :: Either a b -> b
```

Comment:
This `fromRight'` function is a blatant hack because I had no idea how to use `fromRight` (from `Data.Either`) since it required a default.
Also, I don't have any idea how to raise an error of using `VariableNotFound`.